src/security_utilities_rust/src/microsoft_security_utilities_core/identifiable_secrets.rs

```diff
@@ -37,6 +50,186 @@ pub fn is_base64_url_encoding_char(ch: char) -> bool
+    actual_checksum_bytes[0] == computed_marvin_bytes[0]
+    && actual_checksum_bytes[1] == computed_marvin_bytes[1]
+    && actual_checksum_bytes[2] == computed_marvin_bytes[2]
+}
+
+pub fn generate_common_annotated_key(base64_encoded_signature: &str,
+    customer_managed_key: bool,
+    platform_reserved: Option<&[u8]>,
+    provider_reserved: Option<&[u8]>,
+    test_char: Option<char>) -> Result<String, String> {
+generate_common_annotated_test_key(VERSION_TWO_CHECKSUM_SEED.clone(),
+      base64_encoded_signature,
+      customer_managed_key,
+      platform_reserved,
+      provider_reserved,
+      test_char)
+}
+
+pub fn generate_common_annotated_test_key(
+    checksum_seed: u64,
+    base64_encoded_signature: &str,
+    customer_managed_key: bool,
+    platform_reserved: Option<&[u8]>,
+    provider_reserved: Option<&[u8]>,
+    test_char: Option<char>,
+) -> Result<String, String> {
+    const PLATFORM_RESERVED_LENGTH: usize = 9;
+    const PROVIDER_RESERVED_LENGTH: usize = 3;
+
+    let platform_reserved = match platform_reserved {
+        Some(reserved) if reserved.len() != PLATFORM_RESERVED_LENGTH => {
+            return Err(format!(
+                "When provided, there must be {} reserved bytes for platform metadata.",
+                PLATFORM_RESERVED_LENGTH
+            ));
+        }
+        Some(reserved) => reserved,
+        None => &[0; PLATFORM_RESERVED_LENGTH],
+    };
+
+    let provider_reserved = match provider_reserved {
+        Some(reserved) if reserved.len() != PROVIDER_RESERVED_LENGTH => {
+            return Err(format!(
+                "When provided, there must be {} reserved bytes for resource provider metadata.",
+                PROVIDER_RESERVED_LENGTH
+            ));
+        }
+        Some(reserved) => reserved,
+        None => &[0; PROVIDER_RESERVED_LENGTH],
+    };
+
+    let base64_encoded_signature = if customer_managed_key {
+        base64_encoded_signature.to_uppercase()
+    } else {
+        base64_encoded_signature.to_lowercase()
+    };
+
+    let mut key = String::new();
+
+    loop {
+        let key_length_in_bytes = 66;
+        let mut key_bytes = vec![0; key_length_in_bytes];
+
+        if let Some(test_char) = test_char {
+            key = format!("{:86}==", test_char.to_string().repeat(86));
+        } else {
+            let mut rng = rand::thread_rng();
+            rng.fill_bytes(&mut key_bytes);
+        }
+
+        let key_string = general_purpose::STANDARD.encode(&key_bytes);
+        key_bytes = general_purpose::STANDARD.decode(&key_string).unwrap();
+
+        let j_bits = b'J' - b'A';
+        let q_bits = b'Q' - b'A';
+
+        let reserved = (j_bits as i32) << 18 | (q_bits as i32) << 12 | (q_bits as i32) << 6 | j_bits as i32;
+        let reserved_bytes = reserved.to_ne_bytes();
+
+        let key_bytes_length = key_bytes.len();
+
+        key_bytes[key_bytes_length - 27] = reserved_bytes[2];
```

Comment:
@michaelcfanning Since the `Convert.FromBase64String` method actually drops the byte array length to 64 from 66, the indexed values differ from the Rust version.